### ISSUE TITLE
R keras somewhat slower than python keras 

### ISSUE THREAD

**Issue by szilard** [rstudio/keras#29]
Hi @jjallaire, first of all thanks for awesome R keras package, now R will be on par with Python for deep learning :)

I've been toying around and it seems the "same" code runs slower in R. On the MNIST dataset about 5sec/epoch for R and 3sec/epoch for Python with TF backend on Tesla K80 GPUs on AWS p2.xlarge instance. 

I would think once training started, runtime should be the same, and R profiler shows all time is spent in `.Call`. Any ideas why the difference?

R code [here](https://github.com/szilard/benchm-dl/blob/master/cnn-mnist/1z-keras.R)
Python code [here](https://github.com/szilard/benchm-dl/blob/master/cnn-mnist/1-keras.py)

**Comment by jjallaire** [comment]
Good one! On the same instance type w/ GPU I get 55 seconds for R and 40 seconds for Python. I have one pretty good theory about what might be going on (it has to do with how we convert R matrices to NumPy arrays and if I'm correct it will be straightforward to resolve). More soon!

**Comment by szilard** [comment]
Cool, thanks for looking into it.

**Comment by jjallaire** [comment]
Thanks again for reporting this, fix is here: https://github.com/rstudio/keras/commit/1d6c9f2da02ed3c6a0f64e0e1bcbd4e892a8baee

The problem was caused by the fact that when R matrices and arrays are marshaled to Python we convert them to NumPy arrays that map directly to the R managed memory (i.e. no copy is made and NumPy just uses the already allocated R array/matrix). This implies that the NumPy array will be Fortran-ordered (since that's the default for R) rather than C-ordered (the default for NumPy).

In the case of Keras training, between shuffling, drawing batches, copying to the GPU, and 10 epochs there are a ton of accesses to the Fortran-ordered array which add up to a ~ 40% performance penalty for training this dataset. 

This change ensures that we do a single re-ordering of arrays/matrices to C-column order before we pass them to Keras, which eliminates the penalty we were paying for the accesses to the Fortran-ordered array.

@bwlewis Do you think the reticulate behavior of preserving R Fortran-ordered array memory when marshaling to NumPy is an anti-pattern? Alternatively it could be that this particular scenario of hundreds of accesses + GPU copying is problematic but that for most cases it's a performance win. Let me know what you think.

**Comment by szilard** [comment]
@jjallaire Nice :)
